dev/AppNotifications/ShellLocalization.cpp

```diff
@@ -0,0 +1,272 @@
+    THROW_IF_FAILED(spImageStream->Seek(lnBeginning, STREAM_SEEK_SET, nullptr /* new seek pointer */));
+
+    return spImageStream;
+}
+
+void SaveImageWithWIC(
+    winrt::com_ptr<IWICImagingFactory> const& wicImagingFactory,
+    winrt::com_ptr<IWICBitmapSource> const& wicBitmapSource,
+    GUID const& guidContainerFormat,
+    winrt::com_ptr<IStream>& pStream)
+{
+    winrt::com_ptr<IStream> spImageStream = GetStreamOfWICBitmapSource(wicImagingFactory, wicBitmapSource, guidContainerFormat, BITMAP_VERSION::VERSION1);
+
+    // Seek the stream to the beginning and transfer
+    LARGE_INTEGER const lnBeginning{ 0 };
+    THROW_IF_FAILED(spImageStream->Seek(lnBeginning, STREAM_SEEK_SET, nullptr /* new seek pointer */));
+
+    ULARGE_INTEGER lnbuffer{ INT_MAX };
+    THROW_IF_FAILED(spImageStream->CopyTo(pStream.get(), lnbuffer, nullptr /* pointer to number of bytes read */, nullptr /* pointer to number of bytes written */));
+}
+
+void WriteHIconToPngFile(wil::unique_hicon const& hIcon, _In_ PCWSTR pszFileName)
+{
+    auto wicImagingFactory{ winrt::create_instance<IWICImagingFactory>(CLSID_WICImagingFactory, CLSCTX_INPROC_SERVER) };
+
+    winrt::com_ptr<IWICBitmap> wicBitmap;
+    THROW_IF_FAILED(wicImagingFactory->CreateBitmapFromHICON(hIcon.get(), wicBitmap.put()));
+
+    // Create stream to save the HICON to.
+    winrt::com_ptr<IStream> spStream;
+    THROW_IF_FAILED(CreateStreamOnHGlobal(nullptr /* handle */, TRUE, spStream.put()));
+
+    winrt::com_ptr<IWICBitmapSource> wicBitmapSource;
+    wicBitmapSource = wicBitmap;
+
+    SaveImageWithWIC(wicImagingFactory, wicBitmapSource, GUID_ContainerFormatPng, spStream);
+
+    // Write out the stream to a file.
+    winrt::com_ptr<IStream> spStreamOut;
+    THROW_IF_FAILED(SHCreateStreamOnFileEx(pszFileName, STGM_CREATE | STGM_WRITE | STGM_SHARE_DENY_WRITE, FILE_ATTRIBUTE_NORMAL, TRUE, nullptr, spStreamOut.put()));
+
+    STATSTG statstg{};
+    THROW_IF_FAILED(spStream->Stat(&statstg, STATFLAG_NONAME));
+
+    THROW_IF_FAILED(IStream_Reset(spStream.get()));
+
+    THROW_IF_FAILED(spStreamOut->SetSize(statstg.cbSize));
+
+    THROW_IF_FAILED(spStream->CopyTo(spStreamOut.get(), statstg.cbSize, nullptr /* pointer to number of bytes read */, nullptr /* pointer to number of bytes written */));
+
+    THROW_IF_FAILED(spStreamOut->Commit(STGC_DANGEROUSLYCOMMITMERELYTODISKCACHE));
+}
+
+HRESULT IsIconFileExtensionSupported(std::filesystem::path const& iconFilePath)
+{
+    const auto fileExtension = iconFilePath.extension();
+
+    std::string lowercaseFileExtension{ fileExtension.u8string() };
+
+    std::transform(lowercaseFileExtension.begin(), lowercaseFileExtension.end(), lowercaseFileExtension.begin(),
+        [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
+
+    const bool isFileExtensionSupported =
+        lowercaseFileExtension == ".ico" || lowercaseFileExtension == ".bmp" || lowercaseFileExtension == ".jpg" || lowercaseFileExtension == ".png";
+
```

Comment:
Instead of converting ( copying) the string to UTF8 and then lowercase (and that's doing a lexical conversion) better to use CompareStringOrdinal

```c++
static PCWSTR c_supportedExtensions[]{ L".bmp", L".ico", L".jpg", L".png" };
...
const auto fileExtension{ iconFilePath.extension() };
const auto extension{ fileExtension.c_str() };
for (var supportedExtension : c_supportedExtensions)
{
    if (CompareStringOrdinal(extension, -1, supportedExtension, -1, TRUE) == CSTR_EQUAL)
    {
        return true;
    }
}
return true;
```